### PR TITLE
[WiP] DNS/TLS scripts

### PR DESCRIPTION
**Pull request by michielbdejong**
Scripts to let the Box talk to the [DNS server](https://github.com/fxbox/dns-server).

### CODE REVIEW

File: scripts/getUrl.js [hunk 1]
```diff
@@ -0,0 +1,2 @@
+var process = require('process');
+console.log('http://ns.useraddress.net:5300/v1/dns/' + process.argv[2].split('.').reverse().join('/') + '/_acme-challenge');
```
Comment (samgiles) [inline scripts/getUrl.js]: We can replace this file with inline perl :)

`perl -lpe '$_ = join "/", reverse split /\./'`

Comment (michielbdejong) [inline scripts/getUrl.js]: Thanks! Will do.

Comment (fabricedesre) [inline scripts/getUrl.js]: So we will need node.js + perl just to run that getUrl.js script?

Comment (michielbdejong) [inline scripts/getUrl.js]: No, perl will replace node.js.

File: scripts/update.sh [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+node ./apiCall.js knilxof.org 5300 /v1/dns/org/knilxof/box/$1/a "{\"type\":\"A\",\"value\":\"$2\"}"
+node ./apiCall.js knilxof.org 5300 /v1/dns/org/knilxof/box/$1/b "{\"type\":\"A\",\"value\":\"$2\"}"
+node ./apiCall.js knilxof.org 5300 /v1/dns/org/knilxof/box/$1/remote "{\"type\":\"A\",\"value\":\"$2\"}"
+
+echo "$1.box.knilxof.org a.$1.box.knilxof.org b.$1.box.knilxof.org remote.$1.box.knilxof.org" > ./domains.txt
+echo "Getting SAN cert for: `cat domains.txt`"
+./letsencrypt.sh --cron --challenge dns-01 --hook ./deploy-challenge.sh
+
+echo "Setting remote. to use the tunnel"
+node ./apiCall.js knilxof.org 5300 /v1/dns/org/knilxof/box/$1/remote "{\"type\":\"A\",\"value\":\"$3\"}"
```
Comment (michielbdejong) [inline scripts/update.sh:17]: I'm thinking now we can do this already on line 10, not sure what I was thinking here. ;)

File: scripts/apiCall.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+var args = process.argv;
+var options = {
+  hostname: args[2],
+  port: args[3],
+  path: args[4],
+  method: 'POST',
+  key: fs.readFileSync('certs/server/my-server.key.pem'),
+  cert: fs.readFileSync('certs/server/my-server.crt.pem'),
+  rejectUnauthorized: false
```
Comment (michielbdejong) [inline scripts/apiCall.js:12]: You can get rid of this, I think. https://knilxof.org:5300/ now uses a valid cert.

File: scripts/apiCall.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  });
+});
+
+// write data to request body
+req.write(args[5]);
+req.end();
+
+req.on('error', (e) => {
+  console.error(e);
+});
```
Comment (michielbdejong) [inline scripts/apiCall.js:30]: To get rid of this nodejs script, we can use curl:

``` bash
openssl pkcs12 -export -out p.pfx -inkey scripts/certs/server/my-server.key.pem \
  -in scripts/certs/server/my-server.crt.pem 
# enter password 'asdfasdf' twice
curl -k -v --cert p.pfx:asdfasdf -X POST -d'{"type":"A","value":"1.1.1.1"}' \
  https://knilxof.org:5300/v1/dns/org/knilxof/box/315350997cbf4cd4aed2cf2eddd73e8691ab24da
dig 315350997cbf4cd4aed2cf2eddd73e8691ab24da.box.knilxof.org
```

**Comment by samgiles** [comment]
I've written _most_ of the certificate generation in Rust, however creating a CA signed certificate is currently not possible with the rust openssl bindings: https://github.com/sfackler/rust-openssl/issues/345.

**Comment by michielbdejong** [comment]
> I've written most of the certificate generation in Rust

Do you mean including `scripts/letsencrypt.sh`? Why would we want to translate that to Rust? It's a bash script that needs to be called (e.g. once a day or once a month) from a cron job. I would prefer to stay up to date with [the upstream](https://github.com/lukas2511/letsencrypt.sh)?

**Comment by samgiles** [comment]
No, everything but `letsencrypt.sh`

**Comment by ferjm** [comment]
@michielbdejong can we close this PR?

**Comment by michielbdejong** [comment]
Yes

**Pull request closed without merging**